File: app/src/main/java/com/example/lanhuajian/blues/framework/widget/explosionanimator/FallingParticle.java
package com.example.lanhuajian.blues.framework.widget.explosionanimator;

import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.Rect;

public class FallingParticle extends Particle {

    float radius = FallingParticleFactory.PART_WH;
    float alpha = 1.0f;
    Rect mRect;

    public FallingParticle(int color, float cx, float cy, Rect rect) {
        super(color, cx, cy);
        mRect = rect;
    }

    @Override
    protected void draw(Canvas canvas, Paint paint) {
        paint.setColor(color);
        paint.setAlpha((int) (Color.alpha(color) * alpha));
        canvas.drawCircle(cx, cy, radius, paint);

    }

    @Override
    protected void calculate(float factor) {
        cx += factor * BitmapUtils.RANDOM.nextInt(mRect.width()) * ((BitmapUtils.RANDOM.nextFloat() - 0.5f));
        cy += factor * BitmapUtils.RANDOM.nextInt(mRect.height() / 2);
        radius -= factor * BitmapUtils.RANDOM.nextInt(2);
        alpha = (1f - factor) * (BitmapUtils.RANDOM.nextFloat() + 1);
    }
}

File: app/src/main/java/com/blues/framework/widget/explosionanimator/FallingParticle.kt
package com.blues.framework.widget.explosionanimator

import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.graphics.Rect

class FallingParticle(color: Int, cx: Float, cy: Float, var mRect: Rect) : Particle(color, cx, cy) {

    var radius = FallingParticleFactory.PART_WH.toFloat()
    var alpha = 1.0f
    override fun draw(canvas: Canvas, paint: Paint) {
        paint.color = color
        paint.alpha = (Color.alpha(color) * alpha).toInt()
        canvas.drawCircle(cx, cy, radius, paint)
    }

    override fun calculate(factor: Float) {
        cx += factor * BitmapUtils.RANDOM.nextInt(
                mRect.width()) * (BitmapUtils.RANDOM.nextFloat() - 0.5f)
        cy += factor * BitmapUtils.RANDOM.nextInt(mRect.height() / 2)
        radius -= factor * BitmapUtils.RANDOM.nextInt(2)
        alpha = (1f - factor) * (BitmapUtils.RANDOM.nextFloat() + 1)
    }
}